src/query/storages/fuse/src/io/read/agg_index/agg_index_reader_native.rs

```diff
@@ -12,62 +12,158 @@
+                    self.compression.into(),
+                    None,
+                    None,
+                    None,
+                );
+                let res = self
+                    .reader
+                    .async_read_native_columns_data(part)
+                    .await
+                    .inspect_err(|e| debug!("Read aggregating index `{loc}` failed: {e}"))
+                    .ok()?;
+                Some(res)
+            }
+            Err(e) => {
+                if e.kind() == opendal::ErrorKind::NotFound {
+                    debug!("Aggregating index `{loc}` not found.")
+                } else {
+                    debug!("Read aggregating index `{loc}` failed: {e}");
+                }
+                None
             }
-            let is_nested = !nread::reader::is_primitive(field.data_type());
-            let array = nread::batch_read::batch_read_array(
-                readers,
-                curr_leaves,
-                field.clone(),
-                is_nested,
-                pages,
-            )?;
-            arrays.push(array);
         }
-        let chunk = Chunk::new(arrays);
-        let block = DataBlock::from_arrow_chunk(&chunk, &self.schema)?;
+    }
 
-        // Remove unused columns.
-        let block = DataBlock::resort(
-            block,
-            &self.schema,
-            &self.reader.projected_schema.as_ref().into(),
-        )?;
+    pub fn deserialize_native_data(&self, data: &mut NativeSourceData) -> Result<DataBlock> {
+        let mut all_columns_arrays = vec![];
+        for (index, column_node) in self.reader.project_column_nodes.iter().enumerate() {
+            let column_leaves = column_node
+                .leaf_indices
+                .iter()
+                .map(|i| self.reader.parquet_schema_descriptor.columns()[*i].clone())
+                .collect::<Vec<_>>();
 
+            let readers = data.remove(&index).unwrap();
+            let array_iter = BlockReader::build_array_iter(column_node, column_leaves, readers)?;
+            let arrays = array_iter.map(|a| Ok(a?)).collect::<Result<Vec<_>>>()?;
+            all_columns_arrays.push(arrays);
+        }
+        if all_columns_arrays.is_empty() {
+            return Ok(DataBlock::empty_with_schema(Arc::new(
+                self.reader.data_schema(),
+            )));
+        }
+        debug_assert!(
+            all_columns_arrays
+                .iter()
+                .all(|a| a.len() == all_columns_arrays[0].len())
+        );
+        let page_num = all_columns_arrays[0].len();
+        let mut blocks = Vec::with_capacity(page_num);
+
+        for i in 0..page_num {
+            let mut arrays = Vec::with_capacity(all_columns_arrays.len());
+            for array in all_columns_arrays.iter() {
+                arrays.push(array[i].clone());
+            }
+            let chunk = Chunk::new(arrays);
+            let block = DataBlock::from_arrow_chunk(&chunk, &self.reader.data_schema())?;
```

Comment:
There is `batch_read_array` api in strawboat.

https://github.com/sundy-li/strawboat/blob/71acd5250c664603b51767a5b89f5515cd49753a/examples/strawboat_batch_read.rs

Reply:
`batch_read_array ` needs a raw reader, but we get a native reader with page meta.